src/js_parser.zig

```diff
@@ -23102,677 +23059,355 @@ fn NewParser_(
-                        if (p.has_import_meta) {
-                            p.import_meta_ref = p.newSymbol(.other, "$Bun_import_meta") catch bun.outOfMemory();
-                            args[5] = Arg{ .binding = p.b(B.Identifier{ .ref = p.import_meta_ref }, logger.Loc.Empty) };
-                        }
+            // _s(func, "<hash>", force, () => [useCustom])
+            return p.newExpr(E.Call{
+                .target = Expr.initIdentifier(ctx.signature_cb, loc),
+                .args = ExprNodeList.init(args),
+            }, loc);
+        }
 
-                        var total_stmts_count: usize = 0;
-                        for (parts) |part| {
-                            total_stmts_count += part.stmts.len;
-                        }
+        pub fn toAST(
+            p: *P,
+            input_parts: []js_ast.Part,
+            exports_kind: js_ast.ExportsKind,
+            wrap_mode: WrapMode,
+            hashbang: []const u8,
+        ) !js_ast.Ast {
+            const allocator = p.allocator;
+            var parts = input_parts;
 
-                        const preserve_strict_mode = p.module_scope.strict_mode == .explicit_strict_mode and
-                            !(parts.len > 0 and
-                            parts[0].stmts.len > 0 and
-                            parts[0].stmts[0].data == .s_directive);
+            // if (p.options.tree_shaking and p.options.features.trim_unused_imports) {
+            //     p.treeShake(&parts, false);
+            // }
 
-                        total_stmts_count += @as(usize, @intCast(@intFromBool(preserve_strict_mode)));
+            const bundling = p.options.bundle;
+            var parts_end: usize = @as(usize, @intFromBool(bundling));
 
-                        const stmts_to_copy = allocator.alloc(Stmt, total_stmts_count) catch bun.outOfMemory();
-                        {
-                            var remaining_stmts = stmts_to_copy;
-                            if (preserve_strict_mode) {
-                                remaining_stmts[0] = p.s(
-                                    S.Directive{
-                                        .value = "use strict",
-                                    },
-                                    p.module_scope_directive_loc,
-                                );
-                                remaining_stmts = remaining_stmts[1..];
-                            }
+            // When bundling with HMR, we need every module to be just a
+            // single part, as we later wrap each module into a function,
+            // which requires a single part. Otherwise, you'll end up with
+            // multiple instances of a module, each with different parts of
+            // the file. That is also why tree-shaking is disabled.
+            if (p.options.features.hot_module_reloading) {
+                bun.assert(!p.options.tree_shaking);
+                bun.assert(p.options.features.hot_module_reloading);
```

Comment:
it's harmless but this assert isn't necessary

Reply:
good catch